pkg/langmgr/langmgr.go

```diff
@@ -0,0 +1,159 @@
+package langmgr
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/moby/buildkit/client/llb"
+	"github.com/project-copacetic/copacetic/pkg/buildkit"
+	"github.com/project-copacetic/copacetic/pkg/types/unversioned"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	copaPrefix     = "copa-"
+	resultsPath    = "/" + copaPrefix + "out"
+	downloadPath   = "/" + copaPrefix + "downloads"
+	unpackPath     = "/" + copaPrefix + "unpacked"
+	resultManifest = "langresults.manifest"
+)
+
+type LangManager interface {
+	InstallUpdates(context.Context, *llb.State, *unversioned.UpdateManifest, bool) (*llb.State, []string, error)
+}
+
+// GetLanguageManagers returns a list of all available language managers.
+// As new language managers are implemented, they should be added to this list.
+func GetLanguageManagers(config *buildkit.Config, workingFolder string) []LangManager {
+	var managers []LangManager
+
+	// Add Python manager
+	managers = append(managers, &pythonManager{config: config, workingFolder: workingFolder})
+
+	return managers
+}
+
+// Utility functions for package manager implementations to share
+
+type VersionComparer struct {
+	IsValid  func(string) bool
+	LessThan func(string, string) bool
+}
+
+func GetUniqueLatestUpdates(
+	updates unversioned.LangUpdatePackages,
+	cmp VersionComparer,
+	ignoreErrors bool,
+) (unversioned.LangUpdatePackages, error) {
+	if len(updates) == 0 {
+		return unversioned.LangUpdatePackages{}, nil
+	}
+
+	dict := make(map[string]string)
+	var allErrors *multierror.Error
+	for _, u := range updates {
+		switch {
+		case u.FixedVersion == "":
+			// No suitable version found due to patch level restrictions, skip this update
+			log.Debugf("Skipping package %s: no suitable version found according to patch level restrictions", u.Name)
+			continue
+		case cmp.IsValid(u.FixedVersion):
+			ver, ok := dict[u.Name]
+			if !ok {
+				dict[u.Name] = u.FixedVersion
+			} else if cmp.LessThan(ver, u.FixedVersion) {
+				dict[u.Name] = u.FixedVersion
+			}
+		default:
+			err := fmt.Errorf("invalid version %s found for package %s", u.FixedVersion, u.Name)
+			log.Error(err)
+			allErrors = multierror.Append(allErrors, err)
+			continue
+		}
+	}
+	if allErrors != nil && !ignoreErrors {
+		return nil, allErrors.ErrorOrNil()
+	}
+
+	out := unversioned.LangUpdatePackages{}
+	for k, v := range dict {
+		out = append(out, unversioned.UpdatePackage{Name: k, FixedVersion: v})
+	}
+	return out, nil
+}
+
+type UpdatePackageInfo struct {
+	Filename string
+	Version  string
+}
+
+type PackageInfoReader interface {
+	GetVersion(string) (string, error)
+	GetName(string) (string, error)
+}
+
+type UpdateMap map[string]*UpdatePackageInfo
+
+func GetValidatedUpdatesMap(
+	updates unversioned.LangUpdatePackages,
```

Comment:
since the functionality for validating updates and version comparing is similar with what we do for pkgmgr, maybe we could combine this into a helper folder to share the code?

Reply:
can we do a follow up for this one